File: tests/AppConfigTests.cs
// =============================================================================
// CyberPatriot Automation Tool - AppConfig Tests
// Author: Maxwell McCormick
// Copyright (c) 2026 Maxwell McCormick. All Rights Reserved.
// =============================================================================

using CyberPatriotAutomation.Core;
using FluentAssertions;
using Xunit;

namespace CyberPatriotAutomation.Tests;

/// <summary>
/// Unit tests for AppConfig
/// </summary>
public class AppConfigTests
{
    [Fact]
    public void Version_ShouldBeDefined()
    {
        AppConfig.Version.Should().NotBeNullOrEmpty();
    }

    [Fact]
    public void CCSClientServiceName_ShouldBeDefined()
    {
        AppConfig.CCSClientServiceName.Should().NotBeNullOrEmpty();
        AppConfig.CCSClientServiceName.Should().Contain("CCS");
    }

    [Fact]
    public void ScoringReportShortcut_ShouldBeDefined()
    {
        AppConfig.ScoringReportShortcut.Should().NotBeNullOrEmpty();
    }

    [Fact]
    public void DefaultReadmePaths_ShouldNotBeEmpty()
    {
        AppConfig.DefaultReadmePaths.Should().NotBeNullOrEmpty();
    }

    [Fact]
    public void DefaultReadmePaths_ShouldContainCommonLocations()
    {
        var paths = AppConfig.DefaultReadmePaths;

        paths.Should().Contain(p => p.Contains("Desktop"));
    }

    [Fact]
    public void SecurePasswords_ShouldHaveEnoughPasswords()
    {
        AppConfig.SecurePasswords.Should().HaveCountGreaterOrEqualTo(10);
    }

    [Fact]
    public void SecurePasswords_ShouldBeUnique()
    {
        var passwords = AppConfig.SecurePasswords;

        passwords.Should().OnlyHaveUniqueItems();
    }

    [Fact]
    public void SecurePasswords_ShouldMeetComplexityRequirements()
    {
        foreach (var password in AppConfig.SecurePasswords)
        {
            // At least 12 characters
            password.Length.Should().BeGreaterOrEqualTo(12);

            // Has uppercase
            password.Should().MatchRegex("[A-Z]");

            // Has lowercase
            password.Should().MatchRegex("[a-z]");

            // Has digit
            password.Should().MatchRegex("[0-9]");

            // Has special character
            password.Should().MatchRegex("[^a-zA-Z0-9]");
        }
    }
}

File: rust/tests/app_config_tests.rs
// =============================================================================
// PinnacleCyPat - AppConfig Tests
// =============================================================================

use pinnacle_cypat::app_config;

#[test]
fn version_should_be_defined() {
    assert!(!app_config::VERSION.is_empty());
}

#[test]
fn ccs_client_service_name_should_be_defined() {
    assert!(!app_config::CCS_CLIENT_SERVICE_NAME.is_empty());
    assert!(app_config::CCS_CLIENT_SERVICE_NAME.contains("CCS"));
}

#[test]
fn scoring_report_shortcut_should_be_defined() {
    assert!(!app_config::SCORING_REPORT_SHORTCUT.is_empty());
}

#[test]
fn default_readme_paths_should_not_be_empty() {
    assert!(!app_config::default_readme_paths().is_empty());
}

#[test]
fn default_readme_paths_should_contain_common_locations() {
    let paths = app_config::default_readme_paths();
    assert!(paths.iter().any(|p| p.contains("Desktop")));
}

// CyberPatriot images conventionally place a `.lnk` shortcut to the README on
// the desktop of the user running the tool, rather than the file itself, so
// `find_readme_file` must resolve shortcuts (via WScript.Shell on Windows)
// before falling back to the literal default paths. This just exercises the
// async entry point end-to-end without asserting a specific outcome, since
// what (if anything) is found depends on the machine running the test.
#[tokio::test]
async fn find_readme_file_does_not_panic() {
    let _ = app_config::find_readme_file().await;
}

#[test]
fn secure_passwords_should_have_enough_passwords() {
    assert!(app_config::SECURE_PASSWORDS.len() >= 10);
}

#[test]
fn secure_passwords_should_be_unique() {
    let mut seen = std::collections::HashSet::new();
    for pw in app_config::SECURE_PASSWORDS {
        assert!(seen.insert(*pw), "duplicate password: {pw}");
    }
}

#[test]
fn secure_passwords_should_meet_complexity_requirements() {
    for password in app_config::SECURE_PASSWORDS {
        assert!(password.chars().count() >= 12, "password too short: {password}");
        assert!(password.chars().any(|c| c.is_ascii_uppercase()), "no uppercase: {password}");
        assert!(password.chars().any(|c| c.is_ascii_lowercase()), "no lowercase: {password}");
        assert!(password.chars().any(|c| c.is_ascii_digit()), "no digit: {password}");
        assert!(password.chars().any(|c| !c.is_ascii_alphanumeric()), "no special: {password}");
    }
}
